File: Python/cryptography/vigenere_cipher.py
# Vigenere Cipher python code.

def encrypt(msg, key):
    cipher = ""
    n = len(key)
    i = 0
    for l in msg:
        # finding the index of the character
        p = int(ord(l) + ord(key[i % n])) % 26
        # to perform the shift with the key
        cipher += chr(p + ord('A'))
        i += 1
    return cipher


def decrypt(cipher, key):
    msg = ""
    n = len(key)
    j = 0
    for l in cipher:
        p = int(ord(l) - ord(key[j % n]) + 26) % 26
        msg += chr(p + ord('A'))
        j += 1
    return msg

if __name__=='__main__':
    
    # Use uppercase only
    choice = int(input("Choose 1 if you want to ENCRYPT or 2 if you want to DECRYPT: "))

    if(choice == 1):
        msg = input("Enter a plaintext message: ")
        key = input("Enter a keyword: ")

        # encrypting the message
        cipher = encrypt(msg, key)
        print("\nEncrypted message: ", cipher)

    elif(choice == 2):
        cipher = input("Enter a ciphertext: ")
        key = input("Enter a keyword: ")

        # decrypting the ciphar text
        original = decrypt(cipher, key)
        print("\nDecrypted message: ", original)

    else:
        print("\nInvalid choice.")

'''
Sample I/O:

1)
Choose 1 if you want to ENCRYPT or 2 if you want to DECRYPT: 1
Enter a plaintext message: NEOALGO
Enter a keyword: MARS

Encrypted message:  ZEFSXGF

2)
Choose 1 if you want to ENCRYPT or 2 if you want to DECRYPT: 2
Enter a ciphertext: ZEFSXGF
Enter a keyword: MARS

Decrypted message: NEOALGO

'''

File: C/cryptography/Vigenere_Cipher.c
/*Vigenere Cipher Algorithm is a famous cryptographic algorithm.
Vigenère cipher is a method of encrypting alphabetic text by using a series of interwoven Caesar ciphers*/

#include <stdio.h>
#include <string.h>

char key[100];
int x;

void encryption(char message[]) {
	for (int i = 0; i < strlen(message); ++i) {
		int t=(message[i] + key[i%x])%26;
    	message[i]=('A'+t);
	}
}

void decryption(char message[]) {
	for (int i = 0; i < strlen(message); ++i) {
	    int t=(message[i] - key[i%x] +26)%26;
	    message[i]=('A'+t);
	}
}

int main() {
	char message[100];
	printf("Enter Message To Encode: ");
	scanf("%s", message);
	//use only uppercase characters (you can modify the code for lowercase characters)
	printf("\nEnter Key:");
	scanf("%s", key);
	x=strlen(key);

	printf("Generated Key: %s\n", key);
	encryption(message);
	printf("Encrypted Message: %s\n",message);

	decryption(message);
	printf("Decrypted Message: %s",message);
	return 0;
}

/*
OUTPUT
Generated Key: NEOALGO
Encrypted Message: GIGSPXOPXQOOOBTRSOLRUB
Decrypted Message: TESSERACTCODINGNEOALGO
*/
